pkg/metadata/manager.go

```diff
@@ -0,0 +1,404 @@
+		pm.containers[cId] = struct{}{}
+		currContainers[cId] = struct{}{}
+
+		// Container could've been gone, so don't log any errors.
+		cgrpPath, err := m.getContainerCgroupPath(id, cId, pod.Status.QOSClass)
+		if err != nil {
+			log.Debugf("failed to get container metadata for (%s): %v", cId, err)
+			continue
+		}
+		cgrpId, err := m.cgroupsChecker.GetCgroupID(cgrpPath)
+		if err != nil {
+			log.Debugf("failed to get cgroup id for cgroup path (%s): %v", cgrpPath, err)
+			continue
+		}
+		log.Infof("aditi-debug-metadata %s %d", cgrpPath, cgrpId)
+		m.containerMetadataByCgrpId[cgrpId] = &containerMetadata{
+			cgroupId:    cgrpId,
+			cgroupPath:  cgrpPath,
+			parentPodId: id,
+		}
+	}
+	// Clean up any pod's old containers.
+	if oldPod != nil {
+		for _, c := range oldPod.Status.ContainerStatuses {
+			if _, ok := currContainers[c.ContainerID]; !ok {
+				delete(pm.containers, c.ContainerID)
+			}
+		}
+	}
+}
+
+func (m *Manager) dumpFiles(iskind bool) {
+	dir := defaults.DefaultCgroupRoot
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	for _, file := range files {
+		name := file.Name()
+		if strings.Contains(name, "kubepods") || strings.Contains(name, "pod") {
+			log.Infof("aditi-debug %s", name)
+		} else {
+			log.Infof("others %s", name)
+		}
+	}
+}
+
+func (m *Manager) deletePodMetadata(pod *v1.Pod) {
+	id := string(pod.ObjectMeta.UID)
+
+	if _, ok := m.podMetadataById[id]; !ok {
+		return
+	}
+	for k, cm := range m.containerMetadataByCgrpId {
+		if cm.parentPodId == id {
+			delete(m.containerMetadataByCgrpId, k)
+		}
+	}
+	delete(m.podMetadataById, id)
+}
+
+func (m *Manager) getParentPodMetadata(cgroupId uint64, podMetadataOut chan *PodMetadata) {
+	cm, ok := m.containerMetadataByCgrpId[cgroupId]
+	if !ok {
+		log.Debugf("Metadata not found for container: %d", cgroupId)
+		close(podMetadataOut)
+		return
+	}
+
+	pm, ok := m.podMetadataById[cm.parentPodId]
+	if !ok {
+		log.Debugf("Parent pod metadata not found for container: %d", cgroupId)
+		close(podMetadataOut)
+		return
+	}
+	podMetadata := PodMetadata{
+		name:      pm.name,
+		namespace: pm.namespace,
+	}
+	podMetadata.ips = append(podMetadata.ips, pm.ips...)
+	log.Debugf("Parent pod metadata for container (%d): %+v", cgroupId, podMetadata)
+
+	podMetadataOut <- &podMetadata
+	close(podMetadataOut)
+}
+
+func (m *Manager) baseCgroupPathForQos(path string, qos v1.PodQOSClass) string {
+	if qos == v1.PodQOSGuaranteed {
+		return fmt.Sprintf("%s", path)
+	}
+	return fmt.Sprintf("%s/%s", path, strings.ToLower(string(qos)))
+}
+
+func (m *Manager) getContainerCgroupPath(podId string, containerId string, containerQos v1.PodQOSClass) (string, error) {
+	if m.templateCgroupBasePath == "" {
+		return "", fmt.Errorf("failed to get cgroup path for (%s)", containerId)
+	}
+
+	return fmt.Sprintf("%s/pod%s/%s", m.baseCgroupPathForQos(m.templateCgroupBasePath, containerQos),
+		podId, containerId), nil
+}
+
+func newManagerTest(fs fs, cg cgroup) *Manager {
```

Comment:
nit: Does this test code need to be in the main file? Seems like there could perhaps be a little more sharing with the main NewManager implementation and/or shift this test one that passes in dummy `fs` and `cg` implementations into the test file itself.